controllers/localization_controller.go

```diff
@@ -51,29 +52,74 @@ type LocalizationReconciler struct {
+//     update; the reconciliation should _PROCEED_
+//
+// If neither of these cases match, the reconciliation should _STOP_ and requeue the object.
+// For mutating objects, we include two other checks. These objects can have a Source object defined as a Snapshot.
+// We also want to track the condition of those source objects. If the last seen digest or tag of those Snapshots
+// changed, we should reconcile and see if there is anything new we need to apply.
+
 // SetupWithManager sets up the controller with the Manager.
 func (r *LocalizationReconciler) SetupWithManager(mgr ctrl.Manager) error {
-	snapshotSourceKey := ".metadata.snapshot.source"
-	configKey := ".metadata.config"
+	const (
+		sourceKey      = ".metadata.source"
+		configKey      = ".metadata.config"
+		patchSourceKey = ".metadata.patchSource"
+	)
+
+	if err := mgr.GetFieldIndexer().IndexField(context.TODO(), &v1alpha1.Localization{}, sourceKey, func(rawObj client.Object) []string {
+		loc := rawObj.(*v1alpha1.Localization)
+		return []string{loc.Spec.SourceRef.Name}
 
-	if err := mgr.GetCache().IndexField(context.TODO(), &v1alpha1.Localization{}, snapshotSourceKey,
-		r.indexBy("Snapshot", "SourceRef")); err != nil {
+	}); err != nil {
 		return fmt.Errorf("failed setting index fields: %w", err)
 	}
 
-	if err := mgr.GetCache().IndexField(context.TODO(), &v1alpha1.Localization{}, configKey,
-		r.indexBy("ComponentDescriptor", "ConfigRef")); err != nil {
+	if err := mgr.GetFieldIndexer().IndexField(context.TODO(), &v1alpha1.Localization{}, configKey, func(rawObj client.Object) []string {
+		loc := rawObj.(*v1alpha1.Localization)
+		if loc.Spec.ConfigRef == nil {
+			return nil
+		}
+		return []string{loc.Spec.ConfigRef.Name}
+
+	}); err != nil {
+		return fmt.Errorf("failed setting index fields: %w", err)
+	}
+
+	if err := mgr.GetFieldIndexer().IndexField(context.TODO(), &v1alpha1.Localization{}, patchSourceKey, func(rawObj client.Object) []string {
+		loc := rawObj.(*v1alpha1.Localization)
+		if loc.Spec.ConfigRef == nil {
+			return nil
+		}
+		return []string{loc.Spec.PatchStrategicMerge.Source.SourceRef.Name}
+
+	}); err != nil {
 		return fmt.Errorf("failed setting index fields: %w", err)
 	}
 
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&v1alpha1.Localization{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
+		Watches(
+			&source.Kind{Type: &v1alpha1.ComponentVersion{}},
+			handler.EnqueueRequestsFromMapFunc(r.findObjects(sourceKey, configKey)),
+			builder.WithPredicates(ComponentVersionChangedPredicate{}),
+		).
 		Watches(
 			&source.Kind{Type: &v1alpha1.Snapshot{}},
-			handler.EnqueueRequestsFromMapFunc(r.requestsForRevisionChangeOf(snapshotSourceKey)),
+			handler.EnqueueRequestsFromMapFunc(r.findObjects(sourceKey, configKey)),
+			builder.WithPredicates(SnapshotDigestChangedPredicate{}),
+		).
+		Watches(
+			&source.Kind{Type: &v1beta2.GitRepository{}},
```

Comment:
It feels like we will have to add other sources as well eventually then?

Reply:
Yes that's for sure.